File: android-starter/app/src/main/java/com/evplatform/driver/ui/screens/MapSearchScreen.kt
package com.evplatform.driver.ui.screens

import androidx.compose.animation.AnimatedVisibility
import androidx.compose.foundation.background
import androidx.compose.foundation.layout.*
import androidx.compose.foundation.shape.CircleShape
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.*
import androidx.compose.runtime.*
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import com.evplatform.driver.model.Connector
import com.evplatform.driver.model.Station
import com.evplatform.driver.network.ApiClient
import com.evplatform.driver.ui.components.PlugWatchReportSheet
import com.evplatform.driver.ui.components.TrustBadge
import com.evplatform.driver.ui.theme.EVColor
import com.evplatform.driver.ui.theme.EVRadius
import com.evplatform.driver.ui.theme.ConnectorStatus

/**
 * Map is left to your maps SDK of choice (Google Maps Compose / Mapbox).
 * This screen wires the station list + bottom sheet + start-charging CTA,
 * which is the part specific to this product.
 */
@OptIn(ExperimentalMaterial3Api::class)
@Composable
fun MapSearchScreen(
    stations: List<Station>,
    apiClient: ApiClient,
    onStartCharging: (Station, Connector) -> Unit
) {
    var selected by remember { mutableStateOf<Station?>(stations.firstOrNull()) }
    var reportingConnector by remember { mutableStateOf<Connector?>(null) }

    Box(Modifier.fillMaxSize().background(EVColor.bgDeep)) {
        // TODO: GoogleMap(...) { stations.forEach { StationMarker(it) } }

        AnimatedVisibility(
            visible = selected != null,
            modifier = Modifier.align(Alignment.BottomCenter)
        ) {
            selected?.let { station ->
                StationDetailSheet(
                    station = station,
                    onStart = { connector -> onStartCharging(station, connector) },
                    onReportIssue = { connector -> reportingConnector = connector }
                )
            }
        }
    }

    reportingConnector?.let { connector ->
        ModalBottomSheet(onDismissRequest = { reportingConnector = null }) {
            PlugWatchReportSheet(connector = connector, apiClient = apiClient, onDismiss = { reportingConnector = null })
        }
    }
}

@Composable
private fun StationDetailSheet(station: Station, onStart: (Connector) -> Unit, onReportIssue: (Connector) -> Unit) {
    Column(
        Modifier
            .fillMaxWidth()
            .clip(RoundedCornerShape(topStart = EVRadius.container, topEnd = EVRadius.container))
            .background(EVColor.bgPanel)
            .padding(18.dp)
    ) {
        Row(Modifier.fillMaxWidth(), horizontalArrangement = Arrangement.SpaceBetween) {
            Column {
                Text(station.address.uppercase(), color = EVColor.textMuted, fontSize = 11.sp)
                Text(station.name, color = EVColor.textMain, fontSize = 20.sp, fontWeight = FontWeight.SemiBold)
            }
            ReliabilityPill(station.reliabilityScore)
        }

        Spacer(Modifier.height(12.dp))

        Row(horizontalArrangement = Arrangement.spacedBy(10.dp)) {
            StatBox("₹${station.pricePerKwh.toInt()}", "per kWh")
            StatBox("${station.connectors.maxOf { it.powerKw }.toInt()}kW", "max speed")
            StatBox(
                "${station.connectors.count { it.status == ConnectorStatus.AVAILABLE }}/${station.connectors.size}",
                "free now"
            )
        }

        Spacer(Modifier.height(14.dp))

        station.connectors.forEach { connector -> ConnectorRow(connector, onReportIssue = { onReportIssue(connector) }) }

        Spacer(Modifier.height(14.dp))

        Button(
            onClick = { station.connectors.firstOrNull { it.status == ConnectorStatus.AVAILABLE }?.let(onStart) },
            colors = ButtonDefaults.buttonColors(containerColor = EVColor.lime, contentColor = EVColor.bgDeep),
            shape = EVRadius.capsule,
            modifier = Modifier.fillMaxWidth().height(52.dp)
        ) { Text("Start charging", fontWeight = FontWeight.SemiBold) }
    }
}

@Composable
private fun ReliabilityPill(score: Double) {
    Row(
        verticalAlignment = Alignment.CenterVertically,
        modifier = Modifier.clip(EVRadius.capsule).background(EVColor.bgRaise).padding(horizontal = 10.dp, vertical = 5.dp)
    ) {
        Box(Modifier.size(6.dp).clip(CircleShape).background(EVColor.lime))
        Spacer(Modifier.width(6.dp))
        Text("${score.toInt()}% reliable", color = EVColor.textMuted, fontSize = 11.sp)
    }
}

@Composable
private fun RowScope.StatBox(value: String, label: String) {
    Column(
        horizontalAlignment = Alignment.CenterHorizontally,
        modifier = Modifier.weight(1f).clip(RoundedCornerShape(EVRadius.element)).background(EVColor.bgRaise).padding(10.dp)
    ) {
        Text(value, color = EVColor.textMain, fontWeight = FontWeight.Bold, fontSize = 16.sp)
        Text(label.uppercase(), color = EVColor.textMuted, fontSize = 9.sp)
    }
}

@Composable
private fun ConnectorRow(connector: Connector, onReportIssue: () -> Unit) {
    Column(Modifier.fillMaxWidth().padding(vertical = 10.dp)) {
        Row(
            Modifier.fillMaxWidth(),
            horizontalArrangement = Arrangement.SpaceBetween,
            verticalAlignment = Alignment.CenterVertically
        ) {
            Column {
                Text("${connector.type} · #${connector.ocppConnectorId}", color = EVColor.textMain, fontWeight = FontWeight.Bold, fontSize = 13.sp)
                Text("${connector.powerKw.toInt()} kW", color = EVColor.textMuted, fontSize = 11.sp)
            }
            Column(horizontalAlignment = Alignment.End) {
                Row(verticalAlignment = Alignment.CenterVertically) {
                    Box(Modifier.size(6.dp).clip(CircleShape).background(connector.status.color))
                    Spacer(Modifier.width(6.dp))
                    Text(connector.status.label, color = EVColor.textMuted, fontSize = 11.sp)
                }
                connector.reliabilityScore?.let { score ->
                    Spacer(Modifier.height(6.dp))
                    TrustBadge(reliabilityScore = score, guaranteed = connector.guaranteed)
                }
            }
        }
        TextButton(onClick = onReportIssue, contentPadding = PaddingValues(0.dp)) {
            Text("🚩 Report an issue", color = EVColor.textMuted, fontSize = 11.sp)
        }
    }
}

File: ios-starter/EVCharge/Screens/MapSearchScreen.swift
import SwiftUI
import MapKit

struct MapSearchScreen: View {
    @State private var stations: [Station] = []
    @State private var selectedStation: Station?
    @State private var reportingConnector: Connector?
    @State private var region = MKCoordinateRegion(
        center: CLLocationCoordinate2D(latitude: 12.9352, longitude: 77.6146),
        span: MKCoordinateSpan(latitudeDelta: 0.05, longitudeDelta: 0.05)
    )

    var body: some View {
        ZStack(alignment: .bottom) {
            EVTheme.bgDeep.ignoresSafeArea()

            Map(coordinateRegion: $region, annotationItems: stations) { station in
                MapAnnotation(coordinate: CLLocationCoordinate2D(latitude: station.lat, longitude: station.lng)) {
                    StationPin(station: station, isSelected: station.id == selectedStation?.id)
                        .onTapGesture { withAnimation(.spring()) { selectedStation = station } }
                }
            }
            .ignoresSafeArea(edges: .top)

            if let station = selectedStation {
                StationDetailSheet(station: station, onReportIssue: { reportingConnector = $0 })
                    .transition(.move(edge: .bottom))
            }
        }
        .task {
            // Replace with the driver's real location once CoreLocation auth is wired up.
            stations = (try? await APIClient.shared.searchStations(lat: region.center.latitude, lng: region.center.longitude)) ?? []
        }
        .sheet(item: $reportingConnector) { connector in
            PlugWatchReportSheet(connector: connector)
        }
    }
}

private struct StationPin: View {
    let station: Station
    let isSelected: Bool

    private var worstStatus: ConnectorStatus {
        if station.connectors.contains(where: { $0.status == .available }) { return .available }
        if station.connectors.contains(where: { $0.status == .faulted }) { return .faulted }
        return .occupied
    }

    var body: some View {
        Text("₹\(Int(station.pricePerKwh))")
            .font(EVTheme.Font.mono(10, weight: .bold))
            .foregroundColor(EVTheme.bgDeep)
            .padding(isSelected ? 12 : 8)
            .background(Circle().fill(worstStatus.color))
            .scaleEffect(isSelected ? 1.25 : 1.0)
            .shadow(color: .black.opacity(0.4), radius: 6, y: 3)
    }
}

/// Bottom sheet — pricing, connectors, and the "Start charging" CTA that
/// kicks off APIClient.startSession (idempotent, see Networking/APIClient.swift).
private struct StationDetailSheet: View {
    let station: Station
    var onReportIssue: (Connector) -> Void
    @State private var isStarting = false

    var body: some View {
        VStack(alignment: .leading, spacing: 14) {
            Capsule().fill(EVTheme.line).frame(width: 36, height: 4).frame(maxWidth: .infinity)

            HStack(alignment: .top) {
                VStack(alignment: .leading, spacing: 2) {
                    Text(station.address.uppercased())
                        .font(EVTheme.Font.mono(11)).foregroundColor(EVTheme.textMuted)
                    Text(station.name).font(EVTheme.Font.display(20))
                }
                Spacer()
                ReliabilityPill(score: station.reliabilityScore)
            }

            HStack(spacing: 10) {
                StatBox(value: "₹\(Int(station.pricePerKwh))", label: "per kWh")
                StatBox(value: "\(Int(station.connectors.map(\.powerKw).max() ?? 0))kW", label: "max speed")
                StatBox(value: "\(station.connectors.filter { $0.status == .available }.count)/\(station.connectors.count)", label: "free now")
            }

            ForEach(station.connectors) { connector in
                ConnectorRow(connector: connector, onReportIssue: { onReportIssue(connector) })
            }

            Button {
                Task {
                    isStarting = true
                    defer { isStarting = false }
                    // vehicleId would come from the driver's saved-vehicle selection
                }
            } label: {
                Text(isStarting ? "Starting…" : "Start charging")
                    .font(.system(size: 15, weight: .semibold))
                    .frame(maxWidth: .infinity).padding()
                    .background(EVTheme.lime).foregroundColor(EVTheme.bgDeep)
                    .clipShape(Capsule())
            }
            .disabled(isStarting)
        }
        .padding(18)
        .background(EVTheme.bgPanel)
        .clipShape(RoundedRectangle(cornerRadius: EVTheme.Radius.container, style: .continuous))
    }
}

private struct ReliabilityPill: View {
    let score: Double
    var body: some View {
        HStack(spacing: 6) {
            Circle().fill(EVTheme.lime).frame(width: 6, height: 6)
            Text("\(Int(score))% reliable").font(EVTheme.Font.mono(11))
        }
        .padding(.horizontal, 10).padding(.vertical, 5)
        .background(EVTheme.bgRaise).clipShape(Capsule())
        .foregroundColor(EVTheme.textMuted)
    }
}

private struct StatBox: View {
    let value: String; let label: String
    var body: some View {
        VStack(spacing: 2) {
            Text(value).font(EVTheme.Font.mono(16, weight: .bold)).foregroundColor(EVTheme.textMain)
            Text(label.uppercased()).font(.system(size: 9)).foregroundColor(EVTheme.textMuted)
        }
        .frame(maxWidth: .infinity).padding(10)
        .background(EVTheme.bgRaise).clipShape(RoundedRectangle(cornerRadius: EVTheme.Radius.element))
    }
}

private struct ConnectorRow: View {
    let connector: Connector
    var onReportIssue: () -> Void

    var body: some View {
        VStack(alignment: .leading, spacing: 8) {
            HStack {
                VStack(alignment: .leading, spacing: 2) {
                    Text("\(connector.type) · #\(connector.ocppConnectorId)").font(EVTheme.Font.mono(13, weight: .bold))
                    Text("\(Int(connector.powerKw)) kW").font(EVTheme.Font.mono(11)).foregroundColor(EVTheme.textMuted)
                }
                Spacer()
                VStack(alignment: .trailing, spacing: 6) {
                    HStack(spacing: 6) {
                        Circle().fill(connector.status.color).frame(width: 6, height: 6)
                        Text(connector.status.label).font(EVTheme.Font.mono(11)).foregroundColor(EVTheme.textMuted)
                    }
                    if let score = connector.reliabilityScore {
                        TrustBadge(reliabilityScore: score, guaranteed: connector.guaranteed)
                    }
                }
            }
            Button(action: onReportIssue) {
                Text("🚩 Report an issue").font(EVTheme.Font.mono(11)).foregroundColor(EVTheme.textMuted)
            }
        }
        .padding(.vertical, 10)
        Divider().background(EVTheme.line)
    }
}
